packages/navie/src/lib/apply-context.ts

```diff
@@ -0,0 +1,144 @@
+import { log } from 'console';
+
+import { ContextV2 } from '../context';
+
+const VERBOSE = process.env.DEBUG === 'true';
+
+enum ContextItemStatus {
+  OK = 'ok',
+  SIMILAR_CONTENT = 'similar_content',
+  ITEM_UNDEFINED = 'item_undefined',
+  ITEM_TOO_BIG = 'item_too_big',
+}
+
+export const DEFAULT_SELECTION_FREQUENCY: Map<ContextV2.ContextItemType, number> = Object.entries({
+  [ContextV2.ContextItemType.SequenceDiagram]: 1,
+  [ContextV2.ContextItemType.CodeSnippet]: 5,
+  [ContextV2.ContextItemType.DataRequest]: 2,
+  [ContextV2.ContextItemType.HelpDoc]: 1,
+}).reduce((map, [key, value]) => {
+  map.set(key as ContextV2.ContextItemType, value);
+  return map;
+}, new Map<ContextV2.ContextItemType, number>());
+
+export default function applyContext(
+  context: ContextV2.ContextItem[],
+  characterLimit: number,
+  maxContentLength: number = characterLimit / 5,
+  selectionFrequency: Map<ContextV2.ContextItemType, number> = DEFAULT_SELECTION_FREQUENCY,
+  verbose = VERBOSE
+): ContextV2.ContextItem[] {
+  const itemTypeByIndex = Array<ContextV2.ContextItemType>();
+  for (const [itemType, frequency] of selectionFrequency) {
+    for (let i = 0; i < frequency; i += 1) {
+      itemTypeByIndex.push(itemType);
+    }
+  }
+
+  const availableContext: Array<ContextV2.ContextItem | undefined> = [...context];
+  const roundSize = itemTypeByIndex.length;
+
+  // Select items in a round-robin fashion, to ensure a mix of content types. Heuristically, we
+  // want to see one sequence diagram, N code snippets, and M data requests. If we run out of
+  // one type of content type, we'll continue adding the other types.
+  let index = 0;
+  const selectedContext = new Array<ContextV2.ContextItem>();
+  let availableItemCount = availableContext.length;
+  while (availableItemCount > 0) {
+    const itemType = index % roundSize;
+    index += 1;
+
+    let contextItem: ContextV2.ContextItem | undefined;
+    const itemIndex = availableContext.findIndex(
+      (item) => item?.type === itemTypeByIndex[itemType]
+    );
+    if (itemIndex !== -1) {
+      availableItemCount -= 1;
+      contextItem = availableContext[itemIndex];
+      availableContext[itemIndex] = undefined;
+    }
+
+    if (contextItem) selectedContext.push(contextItem);
+  }
+
+  let charsRemaining = characterLimit;
+  log(`Remaining characters before context: ${charsRemaining}`);
+  const appliedContext = new Array<ContextV2.ContextItem>();
+
+  const itemDescription = (contextItem: ContextV2.ContextItem): string =>
+    [contextItem.type, contextItem.location].filter(Boolean).join(' ');
+
+  const addContextItem = (contextItem: ContextV2.ContextItem): ContextItemStatus => {
+    // Don't consume too much of the character limit on a single item.
+    if (contextItem.content.length > maxContentLength) {
```

Comment:
`charsRemaining` isn't considered, so it appears that this context item could potentially overflow the token limit?

```suggestion
    if (contextItem.content.length > Math.min(maxContentLength, charsRemaining)) {
```

Reply:
It's OK to overflow the token limit somewhat. It's a target to aim for but not expected to be a hard limit.